packages/headless/src/controllers/commerce/context/cart/headless-cart.test.ts

```diff
@@ -1,87 +1,140 @@
+  let engine: MockedCommerceEngine;
+  let options: CartOptions;
   let cart: Cart;
-  let engine: MockCommerceEngine;
 
-  beforeEach(() => {
-    engine = buildMockCommerceEngine();
+  function initEngine(preloadedState = buildMockCommerceState()) {
+    engine = buildMockCommerceEngine(preloadedState);
+  }
+
+  function initCart() {
     cart = buildCart(engine, {options});
+  }
+
+  beforeEach(() => {
+    initEngine();
+    jest.resetAllMocks();
+    options = {
+      items: [
+        {
+          productId: 'product-id-1',
+          quantity: 2,
+          name: 'product-name-1',
+          price: 100,
+        },
+        {
+          productId: 'product-id-2',
+          quantity: 4,
+          name: 'product-name-2',
+          price: 25,
+        },
+      ],
+    };
+    initCart();
   });
 
-  const expectContainAction = (action: Action) => {
-    const found = engine.actions.find((a) => a.type === action.type);
-    expect(engine.actions).toContainEqual(found);
-  };
-
-  const expectContainActionWithPayload = (
-    action: Action,
-    payload: string | object
-  ) => {
-    expect(engine.actions).toEqual(
-      expect.arrayContaining([
-        expect.objectContaining({
-          type: action.type,
-          payload:
-            typeof payload === 'string'
-              ? expect.stringContaining(payload)
-              : expect.objectContaining(payload),
-        }),
-      ])
-    );
-  };
+  it('initializes', () => {
+    expect(cart).toBeTruthy();
+  });
+
+  it('exposes a #subscribe method', () => {
+    expect(cart.subscribe).toBeTruthy();
+  });
 
   it('adds the correct reducers to engine', () => {
     expect(engine.addReducers).toHaveBeenCalledWith({
       cart: cartReducer,
+      context: contextReducer,
     });
   });
 
-  it('dispatches #setItems on load if items are present', () => {
-    expectContainAction(setItems);
+  it('dispatches #setItems with correct payload if options include items', () => {
+    const mockedSetItems = jest.mocked(setItems);
+
+    expect(mockedSetItems).toHaveBeenCalledWith(options.items);
+    expect(engine.dispatch).toHaveBeenCalledWith(
+      mockedSetItems.mock.results[0].value
+    );
+  });
+
+  it('does not dispatch #setItems if options do not include items', () => {
+    jest.resetAllMocks();
+    const mockedSetItems = jest.mocked(setItems);
+
+    options = {};
+    initCart();
+    expect(mockedSetItems).not.toHaveBeenCalled();
   });
 
-  it('setItems dispatches #setItems', () => {
-    cart.setItems([item]);
-    expectContainActionWithPayload(setItems, [item]);
+  it('#empty calls #updateItem with quantity of 0 for each item in the cart', () => {
+    const updateItemSpy = jest.spyOn(cart, 'updateItem');
+    jest.mocked(itemsSelector).mockReturnValue(options.items!);
+
+    cart.empty();
+
+    const item = options.items![0];
+
+    expect(updateItemSpy).toHaveBeenCalledWith({
+      ...item,
+      quantity: 0,
+    });
   });
```

Comment:
We _could_ group these tests + the `initializes` test case under an `initialization` `describe` group.